src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -119,59 +125,78 @@
      * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
      */
-    private static void cloneAndAnalyzeRepos(
-            Map<RepoLocation, List<RepoConfiguration>> repoLocationMap, String outputPath) throws IOException {
+    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) {
+        Map<RepoLocation, List<RepoConfiguration>> repoLocationMap = groupConfigsByRepoLocation(configs);
         RepoCloner repoCloner = new RepoCloner();
         RepoLocation clonedRepoLocation = null;
 
-        for (RepoLocation location : repoLocationMap.keySet()) {
-            repoCloner.clone(outputPath, repoLocationMap.get(location).get(0));
+        List<RepoLocation> repoLocationList = new ArrayList<>(repoLocationMap.keySet());
+        Set<RepoLocation> successfullyClonedRepos = new HashSet<>();
+
+        // clones another repo in parallel while analyzing a cloned repo (if any)
+        for (RepoLocation location : repoLocationList) {
+            repoCloner.clone(repoLocationMap.get(location).get(0));
 
             if (clonedRepoLocation != null) {
-                analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation),
+                analyzeRepos(outputPath, configs, repoLocationMap.get(clonedRepoLocation),
                         repoCloner.getCurrentRepoDefaultBranch());
+                successfullyClonedRepos.add(clonedRepoLocation);
             }
-            clonedRepoLocation = repoCloner.getClonedRepoLocation(outputPath);
+            clonedRepoLocation = repoCloner.getClonedRepoLocation();
         }
         if (clonedRepoLocation != null) {
-            analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation), repoCloner.getCurrentRepoDefaultBranch());
+            analyzeRepos(outputPath, configs, repoLocationMap.get(clonedRepoLocation),
+                    repoCloner.getCurrentRepoDefaultBranch());
+            successfullyClonedRepos.add(clonedRepoLocation);
         }
+
+        // handles configs that failed to clone and clean up the cloned repo directory.
+        handleCloningFailed(configs, successfullyClonedRepos);
         repoCloner.cleanup();
     }
 
     /**
-     * Analyzes all repos in {@code configs} and generates their report.
+     * Analyzes all repos in {@code configsToAnalyze} and generates their report.
+     * Also removes {@code configsToAnalyze} that failed to analyze from {@code configs}.
      */
-    private static void analyzeRepos(String outputPath, List<RepoConfiguration> configs, String defaultBranch) {
-        for (RepoConfiguration config : configs) {
-            config.updateBranch(defaultBranch);
+    private static void analyzeRepos(String outputPath, List<RepoConfiguration> configs,
+            List<RepoConfiguration> configsToAnalyze, String defaultBranch) {
+        for (RepoConfiguration configToAnalyze : configsToAnalyze) {
```

Comment:
Here is not using an iterator. Using removeIf with for each loop is unsafe as mentioned.

Reply:
My bad. Forgot that I'm still getting the copy of the config by reference.